UaClient/ServiceModel/Ua/CustomEncodingTable.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+using System.Reflection;
+
+namespace Workstation.ServiceModel.Ua
+{
+    public class CustomEncodingTable : IEnumerable<(ExpandedNodeId, Type)>
+    {
+        protected readonly List<(ExpandedNodeId, Type)> encodingTable;
+
+        public CustomEncodingTable()
+        {
+            this.encodingTable = new List<(ExpandedNodeId, Type)>();
+        }
+        
+        public CustomEncodingTable(IEnumerable<(ExpandedNodeId, Type)> encodingTable)
+        {
+            this.encodingTable = encodingTable.ToList();
+        }
+        
+        public void Add(ExpandedNodeId encodingId, Type type)
+        {
+            this.encodingTable.Add((encodingId, type));
+        }
+        
+        public void Add((ExpandedNodeId, Type) item)
+        {
+            this.encodingTable.Add(item);
+        }
+
+        public virtual IEnumerator<(ExpandedNodeId, Type)> GetEnumerator()
+        {
+            return encodingTable.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => this.GetEnumerator();
+    }
+
+    public class CustomEncodingTable<TAttribute> : CustomEncodingTable
+        where TAttribute : Attribute, IEncodingIdAttribute
+    {
+        public CustomEncodingTable()
+        {
+        }
+        
+        public CustomEncodingTable(IEnumerable<(ExpandedNodeId, Type)> encodingTable)
+            : base(encodingTable)
+        {
+        }
+        
+        public CustomEncodingTable(IEnumerable<Type> types)
+        {
+            var table = from type in types
+                        let info = type.GetTypeInfo()
+                        where info.ImplementedInterfaces.Contains(typeof(IEncodable))
+                        let attr = info.GetCustomAttribute<TAttribute>(false)
+                        where attr != null
+                        select (attr.NodeId, type);
+            this.encodingTable.AddRange(table);
+        }
+
+        public CustomEncodingTable(Assembly assembly)
+        {
+            var table = from type in assembly.ExportedTypes
+                        let info = type.GetTypeInfo()
+                        where info.ImplementedInterfaces.Contains(typeof(IEncodable))
+                        let attr = info.GetCustomAttribute<TAttribute>(false)
+                        where attr != null
```

Comment:
Here it is correct, to skip types, because not all types of an assembly must be annotated.

Reply:
I agree we should skip here.